maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt

```diff
@@ -361,12 +367,151 @@ class OfflineMapLayersPickerTest {
+        onView(withRecyclerView(R.id.layers).atPositionOnView(1, R.id.title)).check(matches(withText(file1.name)))
+        onView(withRecyclerView(R.id.layers).atPositionOnView(2, R.id.title)).check(matches(withText(file2.name)))
     }
 
-    private fun launchOfflineMapLayersPicker(): FragmentScenario<OfflineMapLayersPicker> {
+    @Test
+    fun `layers are collapsed by default`() {
+        whenever(referenceLayerRepository.getAll()).thenReturn(
+            listOf(
+                ReferenceLayer("1", TempFiles.createTempFile(), "layer1"),
+                ReferenceLayer("2", TempFiles.createTempFile(), "layer2")
+            )
+        )
+
+        launchFragment()
+
+        scheduler.flush()
+
+        assertLayerCollapsed(1)
+        assertLayerCollapsed(2)
+    }
+
+    @Test
+    fun `recreating maintains expanded layers`() {
+        whenever(referenceLayerRepository.getAll()).thenReturn(listOf(
+            ReferenceLayer("1", TempFiles.createTempFile(), "layer1"),
+            ReferenceLayer("2", TempFiles.createTempFile(), "layer2"),
+            ReferenceLayer("3", TempFiles.createTempFile(), "layer3")
+        ))
+
+        val scenario = launchFragment()
+
+        scheduler.flush()
+
+        onView(withRecyclerView(R.id.layers).atPositionOnView(1, R.id.arrow)).perform(click())
+        onView(withId(R.id.layers)).perform(scrollToPosition<RecyclerView.ViewHolder>(3))
+        onView(withRecyclerView(R.id.layers).atPositionOnView(3, R.id.arrow)).perform(click())
+
+        scenario.recreate()
+
+        assertLayerExpanded(1)
+        assertLayerCollapsed(2)
+        assertLayerExpanded(3)
+    }
+
+    @Test
+    fun `correct path is displayed after expanding layers`() {
+        val file1 = TempFiles.createTempFile()
+        val file2 = TempFiles.createTempFile()
+        whenever(referenceLayerRepository.getAll()).thenReturn(listOf(
+            ReferenceLayer("1", file1, "layer1"),
+            ReferenceLayer("2", file2, "layer2")
+        ))
+
+        launchFragment()
+
+        scheduler.flush()
+
+        onView(withRecyclerView(R.id.layers).atPositionOnView(1, R.id.arrow)).perform(click())
+        onView(withId(R.id.layers)).perform(scrollToPosition<RecyclerView.ViewHolder>(2))
+        onView(withRecyclerView(R.id.layers).atPositionOnView(2, R.id.arrow)).perform(click())
+
+        onView(withRecyclerView(R.id.layers).atPositionOnView(1, R.id.path)).check(matches(withText(file1.absolutePath)))
+        onView(withRecyclerView(R.id.layers).atPositionOnView(2, R.id.path)).check(matches(withText(file2.absolutePath)))
+    }
+
+    @Test
+    fun `clicking delete shows the confirmation dialog`() {
+        whenever(referenceLayerRepository.getAll()).thenReturn(listOf(
+            ReferenceLayer("1", TempFiles.createTempFile(), "layer1")
+        ))
+
+        launchFragment()
+
+        scheduler.flush()
+
+        onView(withRecyclerView(R.id.layers).atPositionOnView(1, R.id.arrow)).perform(click())
```

Comment:
All these `RecyclerView` Espresso assertions and actions make the tests pretty hard to read, and you'll see there are quite a lot of changes if you auto format the file due to the line lengths. Maybe you can pull out some helpers to try and make them a little cleaner. Remember that the result of an `onView` call can be pulled out to a local variable which can save a bunch of repetition and line length. The chained called can also be stacked on new lines like:

```kotlin
onView(...
     .atPositionOnView(...
     .check(...
```

Reply:
I was thinking about that and I'm on it now... if it's the last missing thing please merge and I will add it along with other improvements https://github.com/getodk/collect/issues/6174#issue-2340337219